tutorials/scripting/c_sharp/c_sharp_variant.rst

```diff
@@ -9,17 +9,67 @@ For a detailed explanation of Variant in general, see the :ref:`Variant <class_V
 We recommend avoiding ``Godot.Variant`` unless it is necessary to interact with untyped engine APIs.
 Take advantage of C#'s type safety when possible.
 
-Any of ``Variant.As{TYPE}`` methods or the generic ``Variant.As<T>`` method can be used to convert
-a ``Godot.Variant`` to a C# type. Since the ``Godot.Variant`` type contains implicit conversions
-defined for all the supported types, calling these methods directly is usually not necessary.
+Converting from a supported type to ``Godot.Variant`` can be done using implicit conversions. Also
+available are ``CreateFrom`` method overloads and the generic ``Variant.From<T>`` methods.
 
-Use ``CreateFrom`` method overloads or the generic ``Variant.From<T>`` method to convert a C# type
-to a ``Godot.Variant``.
+.. code-block:: csharp
+
+    int x = 42;
+    Godot.Variant number = x;
+    Godot.Variant hello = "Hello, World!";
+
+    Godot.Variant number2 = Godot.Variant.CreateFrom(x);
+    Godot.Variant number3 = Godot.Variant.From(x);
+
+Implicit conversions to ``Godot.Variant`` make passing variants as method arguments very convenient.
+For example, ``final_val`` of :ref:`tween_property<class_Tween_method_tween_property>`.
+
+.. code-block:: csharp
+
+    Tween tween = CreateTween();
+    tween.TweenProperty(GetNode("Sprite"), "modulate", Colors.Red, 1.0f);
+
+Converting from ``Godot.Variant`` to a supported type can be done using explicit conversions. Also
+available are ``Variant.As{TYPE}`` methods or the generic ``Variant.As<T>`` method.
+
+.. code-block:: csharp
+
+    int cSharpNumber = (int)number;
+    string cSharpHello = (string)hello;
+
+    int cSharpNumber2 = number.As<int>();
+    string cSharpHello2 = hello.AsString();
+
+.. note::
+
+    All methods of converting from ``Godot.Variant`` to a C# type behave the same way. If conversion
+    is not possible, the default value of the target type or an empty array is returned. An
+    exception is not thrown.
+
+A matching type is not necessary for a conversion to succeed. For example, all variants are
+convertible to ``string``.
```

Comment:
In general, converting between Variant types will always "succeed" but I don't think we should be encouraging this.

I guess it depends on what you consider success, the Variant conversion will try to get the "closest" value but that can be confusing in some cases. Here are some interesting examples:

```csharp
Variant v;
int n;

v = "42"; // This is a string Variant.
n = v.As<int>(); // Converts the string to the int '42'.

v = "42.5";
n = v.As<int>(); // Converts the string to the int '42' discarding the decimals.

v = "42a";
n = v.As<int>(); // Converts the string to the int '42' discarding the suffix.
                 // Honestly, I didn't think this would work (I expected '0').

v = "Hello world";
n = v.As<int>(); // Converts the string to the int '0' because it can't convert the text,
                 // but doesn't print any errors.

v = new Vector3(1, 3, 9); // This is a Vector3 Variant.
n = v.As<int>(); // Converts the Vector3 to the int '0' because it can't convert the Vector3,
                 // but doesn't print any errors.

v = 42; // This is a int Variant.
v.As<Vector3>(); // Converts the int to the Vector3 '(0, 0, 0)' because it can't convert the int,
                 // but doesn't print any errors.
```

Reply:
I'm fine cutting it down to just a heads up. I could see this paragraph and example being misconstrued as something to use rather than `ToString`.

> ```cs
> v = "42a";
> n = v.As<int>(); // Converts the string to the int '42' discarding the suffix.
>                  // Honestly, I didn't think this would work (I expected '0').
> ```

It looks like this was documented in 3.0, and it's more than just suffixes: https://docs.godotengine.org/en/3.0/classes/class_int.html, not sure where this text went (if anywhere):

> This method will ignore all non-number characters, so calling int('1e3') will return 13.

It seems like ideally the Godot Engine conversion logic would be documented somewhere, and this page could just say "conversions have the same loose logic as that ->".

Reply:
> It looks like this was documented in 3.0 [...] not sure where this text went

Oh, you're right. It looks like the documentation for the `int(String)` constructor now points to the `String.to_int` method that has this in the description:

> This method removes any non-number character and stops at the first decimal point (`.`)

I agree that ideally all this `Variant` conversion is an engine thing rather than a C#-specific thing so it should probably be documented in the general `Variant` documentation (maybe [here](https://docs.godotengine.org/en/stable/classes/class_variant.html) or [here](https://docs.godotengine.org/en/stable/contributing/development/core_and_modules/variant_class.html)?) and just linked from the C# documentation.

Reply:
I'll leave an exhaustive explanation for later--for now, simplifying this to a scary list.